cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureUtils.java

```diff
@@ -300,41 +302,78 @@ private void deallocateInstance(AuthenticatedContext ac, List<CloudVmInstanceSta
+                        statusesAfterDelete.add(instance);
+                    } else {
+                        vmsNotDeleted.add(instance.getCloudInstance());
+                    }
+        });
+        LOGGER.info("Delete not stopped VMs: {}", vmsNotDeleted.stream().map(CloudInstance::getInstanceId).collect(Collectors.toList()));
+        deleteInstancesAndFillStatuses(ac, statusesAfterDelete, vmsNotDeleted);
+
+        return statusesAfterDelete;
+    }
+
+    private void deleteInstancesAndFillStatuses(AuthenticatedContext ac, List<CloudVmInstanceStatus> statusesAfterDelete, List<CloudInstance> vmsNotDeleted) {
+        try {
+            List<Completable> deleteCompletables = new ArrayList<>();
+            for (CloudInstance vm : vmsNotDeleted) {
+                deleteInstance(ac, statusesAfterDelete, deleteCompletables, vm);
+            }
+            Completable.mergeDelayError(deleteCompletables).await();
+        } catch (CompositeException e) {
+            String errorMessages = e.getExceptions().stream().map(Throwable::getMessage).collect(Collectors.joining());
+            LOGGER.error("Error(s) occured while waiting for vm deletion: {}, statuses: {}", errorMessages, statusesAfterDelete);
+            throw new CloudbreakServiceException("Error(s) occured while waiting for vm deletion: " + errorMessages, e);
+        } catch (RuntimeException e) {
+            LOGGER.error("Error occured while waiting for vm deletion: {}, statuses: {}", e.getMessage(), statusesAfterDelete, e);
+            throw new CloudbreakServiceException("Error occured while waiting for vm deletion: " + e.getMessage(), e);
         }
-        Completable.merge(deleteCompletables).await();
-        return statuses;
+    }
+
+    private void deleteInstance(AuthenticatedContext ac, List<CloudVmInstanceStatus> statusesAfterDelete, List<Completable> deleteCompletables,
+            CloudInstance vm) {
+        String resourceGroupName = azureResourceGroupMetadataProvider.getResourceGroupName(ac.getCloudContext(), vm);
+        AzureClient azureClient = ac.getParameter(AzureClient.class);
+        deleteCompletables.add(azureClient.deleteVirtualMachine(resourceGroupName, vm.getInstanceId())
+                .doOnError(throwable -> {
+                    LOGGER.error("Error happened on azure instance delete: {}", vm, throwable);
+                    statusesAfterDelete.add(new CloudVmInstanceStatus(vm, InstanceStatus.FAILED, throwable.getMessage()));
+                })
+                .doOnCompleted(() -> statusesAfterDelete.add(new CloudVmInstanceStatus(vm, InstanceStatus.TERMINATED)))
+                .subscribeOn(Schedulers.io()));
     }
 
     @Retryable(backoff = @Backoff(delay = 1000, multiplier = 2, maxDelay = 10000), maxAttempts = 5)
     public void deleteCloudResourceInstances(AuthenticatedContext ac, String resourceGroupName, List<String> instanceIdList) {
```

Comment:
it would be worth keeping either this or azureUtils.deleteInstances(ac, vms) only

Reply:
We cannot delete deleteCloudResourceInstances method because it uses instance name, because instance id is missing when the instance creation is in progress. 
On the other hand if we delete deleteInstance method we would lose the extra information about the statuses and the chance to filter out already terminated vms. (Original issue was raised because we hit azure request limit with a 170 node cluster)